lib/src/main/kotlin/com/swmansion/starknet/data/serializers/TransactionSerializer.kt

```diff
@@ -0,0 +1,87 @@
+            is InvokeTransactionV1 -> Json.encodeToJsonElement(InvokeTransactionV1Serializer, value).jsonObject
+            is InvokeTransactionV3 -> Json.encodeToJsonElement(InvokeTransactionV3Serializer, value).jsonObject
+            is DeclareTransactionV2 -> Json.encodeToJsonElement(DeclareTransactionV2Serializer, value).jsonObject
+            is DeclareTransactionV3 -> Json.encodeToJsonElement(DeclareTransactionV3Serializer, value).jsonObject
+            is DeployAccountTransactionV1 -> Json.encodeToJsonElement(DeployAccountTransactionV1Serializer, value).jsonObject
+            is DeployAccountTransactionV3 -> Json.encodeToJsonElement(DeployAccountTransactionV3Serializer, value).jsonObject
+            is DeclareTransactionV0 -> Json.encodeToJsonElement(DeclareTransactionV0.serializer(), value).jsonObject
+            is DeclareTransactionV1 -> Json.encodeToJsonElement(DeclareTransactionV1.serializer(), value).jsonObject
+            is DeployTransaction -> Json.encodeToJsonElement(DeployTransaction.serializer(), value).jsonObject
+            is InvokeTransactionV0 -> Json.encodeToJsonElement(InvokeTransactionV0.serializer(), value).jsonObject
+            is L1HandlerTransaction -> Json.encodeToJsonElement(L1HandlerTransaction.serializer(), value).jsonObject
+        }
+
+        val result = JsonObject(
+            jsonObject.plus("type" to Json.encodeToJsonElement(value.type)),
+        )
+
+        encoder.encodeJsonElement(result)
+    }
+
+    private fun deserializeInvoke(decoder: JsonDecoder, element: JsonElement): InvokeTransaction {
+        val versionElement = element.jsonObject.getOrElse("version") { throw SerializationException("Input element does not contain mandatory field 'version'") }
+
+        val version = decoder.json.decodeFromJsonElement(Felt.serializer(), versionElement)
+        return when (version) {
+            Felt(3) -> decoder.json.decodeFromJsonElement(InvokeTransactionV3.serializer(), element)
+            Felt.ONE -> decoder.json.decodeFromJsonElement(InvokeTransactionV1.serializer(), element)
+            else -> throw IllegalArgumentException("Invalid invoke transaction version '${versionElement.jsonPrimitive.content}'")
+        }
+    }
+
+    private fun deserializeDeclare(decoder: JsonDecoder, element: JsonElement): DeclareTransaction {
+        val versionElement = element.jsonObject.getOrElse("version") { throw SerializationException("Input element does not contain mandatory field 'version'") }
+
+        val version = decoder.json.decodeFromJsonElement(Felt.serializer(), versionElement)
+        return when (version) {
+            Felt(3) -> decoder.json.decodeFromJsonElement(DeclareTransactionV3.serializer(), element)
+            Felt(2) -> decoder.json.decodeFromJsonElement(DeclareTransactionV2.serializer(), element)
+            else -> throw IllegalArgumentException("Invalid declare transaction version '${versionElement.jsonPrimitive.content}'")
```

Comment:
I think we weren't supposed to be able to deserialize declare transaction **payloads** of version `V0` and `V1`. 

But since we're dealing with `Transaction`s here, not `TransactionPayloads`, I think we are supposed to be able to deserialize historic `V1` and `V0` transactions.

Reply:
The same applies to other type-specific deserializers